net/ping/ping.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package ping allows sending ICMP echo requests to a host in order to
+// determine network latency.
+package ping
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"tailscale.com/net/netns"
+	"tailscale.com/types/logger"
+)
+
+type response struct {
+	t   time.Time
+	err error
+}
+
+// Pinger represents a set of ICMP echo requests to be sent at a single time.
+//
+// A new instance should be created for each concurrent set of ping requests;
+// this type should not be reused.
+type Pinger struct {
+	c       net.PacketConn
+	Logf    logger.Logf
+	Verbose bool
+	timeNow func() time.Time
+	id      uint16 // uint16 per RFC 792
+	seq     uint32 // uint16 per RFC 792, but we don't have an atomic function; use uint32
```

Comment:
I might just not use atomics here and put these under the `mu sync.Mutex`. Is it really hot enough to matter?